File: 백준/Silver/2792. 보석 상자/보석 상자.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        int[] arr = new int[m];

        int low = 0, high = 0;

        for (int i = 0; i < m; i++) {
            arr[i] = Integer.parseInt(br.readLine());
            high = Math.max(high, arr[i]);
        }

        int ans = 0;

        while (low <= high) {

            int mid = (low + high) / 2;

            int count = 0;
            for (int i : arr) {
                count += (int) Math.ceil((double) i / mid);
            }

            if (count <= n) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        System.out.println(ans);
    }
}

File: 백준/Silver/2792. 보석 상자/보석 상자.py
import math
import sys
input = sys.stdin.readline

n, m = map(int, input().split())

arr = []

for _ in range(m):
    arr.append(int(input()))

low, high = 1, max(arr)

ans = 0

while low <= high:
    mid = (low + high) // 2

    count = 0
    for i in arr:
        count += math.ceil(i / mid)

    if count <= n:
        ans = mid
        high = mid - 1
    else:
        low = mid + 1

print(ans)